### PR TITLE
Default runtime-id tag

### PR DESCRIPTION
**Pull request by r1viollet**
# What does this PR do?

Provide a default `runtime-id` tag

# Motivation

Ensure that the timeline data can be discovered for native.

# Additional Notes

The `runtime_id` should come from tracing
When we do not find it, we should still generate a default `runtime-id`.

# How to test the change?

I have a minor test to check the format.
I did not add an end to end test for now.

### CODE REVIEW

**Comment by r1viollet** [comment]
image

**Review by sanchda — DISMISSED** [review]
You make this observation in a comment, but expanding on it here for no particular reason.

As far as I know, `runtime-id` is used to fingerprint a precise instance of a service.  The noted strategy works perfectly fine for the other profilers, since they are indeed attached to a single instance of a runtime and they can hook into runtime faculties to change the `runtime-id` if that changes (e.g., atfork handlers).

On the other hand, this concept is slightly problematic for `ddprof` due to the way we track process lineage/lifetime.  The noted procedure is more of a `profiler-id`.  I just want to note this because our concept of what it means to be a `runtime-id` was greatly influenced by how the existing profilers work, but this will absolutely need to be revisited in the future.

My take:
* This solution is fine because it's needed for timeline
* A more durable solution will be computed in the backend by hashing PID, container-id, and host.  Why?  Because this will uniquely identify processes, and that's what matters.
* Instead of PID, a better solution might be to compute something like `comm` name parity.  In other words, if the parent/child PID sequence looks like bash -> bash -> python -> bash -> apache, the change sequence is 0 -> 0 -> 1 -> 2 -> 3 instead of 0 -> 0 -> 1 -> 0 -> 2.  Something like that, at least.  Why?  Because the affected profilers can start in the middle of execution and may be restarted, so having a stable characterization is valuable.

In conclusion:
* This PR is fine, the current situation isn't our fault and the average ddprof user is _basically_ just starting the profiler once anyway
* Some careful thought should be put into adapting this procedure for a profiler with scope of more than one service, and I think that would require some work from the backend.
* LGTM aside from the minor fixups in a separate PR.  Approving now because I have nothing novel to add and I trust those comments will be addressed.
* Thank you!

File: src/uuid.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "uuid.hpp"
+
+namespace dd {
+
+// Copied from dd-trace-dotnet, and previously
+// copied from https://stackoverflow.com/a/60198074
+// We replace std::mt19937 by std::mt19937_64 so we can generate 64bits numbers
+// instead of 32bits
+std::string GenerateUuidV4() {
```
[the remark this answers is not in the document]
Comment (r1viollet) [reply src/uuid.cc]: I wanted to avoid bit-wise operations, so you know have an array of size 32.

**Comment by r1viollet** [comment]
> You make this observation in a comment, but expanding on it here for no particular reason.
> 
> As far as I know, `runtime-id` is used to fingerprint a precise instance of a service. The noted strategy works perfectly fine for the other profilers, since they are indeed attached to a single instance of a runtime and they can hook into runtime faculties to change the `runtime-id` if that changes (e.g., atfork handlers).
> 
> On the other hand, this concept is slightly problematic for `ddprof` due to the way we track process lineage/lifetime. The noted procedure is more of a `profiler-id`. I just want to note this because our concept of what it means to be a `runtime-id` was greatly influenced by how the existing profilers work, but this will absolutely need to be revisited in the future.
> 
> My take:
> 
> * This solution is fine because it's needed for timeline
> * A more durable solution will be computed in the backend by hashing PID, container-id, and host.  Why?  Because this will uniquely identify processes, and that's what matters.
> * Instead of PID, a better solution might be to compute something like `comm` name parity.  In other words, if the parent/child PID sequence looks like bash -> bash -> python -> bash -> apache, the change sequence is 0 -> 0 -> 1 -> 2 -> 3 instead of 0 -> 0 -> 1 -> 0 -> 2.  Something like that, at least.  Why?  Because the affected profilers can start in the middle of execution and may be restarted, so having a stable characterization is valuable.
> 
> In conclusion:
> 
> * This PR is fine, the current situation isn't our fault and the average ddprof user is _basically_ just starting the profiler once anyway
> * Some careful thought should be put into adapting this procedure for a profiler with scope of more than one service, and I think that would require some work from the backend.
> * LGTM aside from the minor fixups in a separate PR.  Approving now because I have nothing novel to add and I trust those comments will be addressed.
> * Thank you!

Thanks for sharing. I agree, this is far from perfect.
It is true that we could focus on the reconciliation within the backend. Though I think there are also cases where there are several runtime-ids for a given PID (thinking of .NET for instance).
There could be a way we can establish some standard IPC mechanisms with the runtimes. We will experiment on this next quarter.

File: include/uuid.hpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+namespace ddprof {
+struct Uuid {
+  Uuid();
+  static constexpr int k_version_position = 12;
+  static constexpr int k_variant_position = 16;
+  static constexpr int k_version = 4;
+  [[nodiscard]] int get_version() const { return data[k_version_position]; }
+  [[nodiscard]] std::string to_string() const;
+  // We could make this smaller, as it is hexadecimal and 32 characters
+  // but we are keeping one byte per element for simplicity
```
Comment (nsavoire) [inline include/uuid.hpp:21]: Not sure how this simplifies things, I would say either store chars and keeps it 32 bytes or store numbers and make it 16 bytes.

**Review by nsavoire — APPROVED** [review]

File: src/uuid.cc [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  data[k_version_position] = k_version;
+  for (int i = 13; i < 16; ++i) {
+    data[i] = dis(gen);
+  }
+  // variant
+  data[16] = dis2(gen);
+
+  for (int i = 17; i < 32; ++i) {
+    data[i] = dis(gen);
+  }
```
[the remark this answers is not in the document]
Comment (r1viollet) [reply src/uuid.cc]: Happy to review the proposal during a live session.

File: include/uuid.hpp (hunk 2 above)
Comment (r1viollet) [reply include/uuid.hpp:21]: I find it easier to reason on the values of every element.
